packet/Manager.py

```diff
@@ -225,3 +259,375 @@ def validate_capacity(self, servers):
+        params = {
+            "deployment_type": deployment_type,
+            "asn": asn,
+            "md5": md5,
+            "use_case": use_case
+        }
+        self.call_api("/projects/%s/bgp-configs" % project_id,
+                      type="POST",
+                      params=params)
+
+    # BGP Session
+    def get_bgp_sessions(self, device_id, params={}):
+        data = self.call_api("/devices/%s/bgp/sessions" % device_id,
+                             type="GET", params=params)
+        bgp_sessions = list()
+        for jsoned in data["bgp_sessions"]:
+            bpg_session = BGPSession(jsoned)
+            bgp_sessions.append(bpg_session)
+        return bgp_sessions
+
+    def create_bgp_session(self, device_id, address_family):
+        data = self.call_api("/devices/%s/bgp/sessions" % device_id,
+                             type="POST",
+                             params={
+                                 "address_family": address_family
+                             })
+        return BGPSession(data)
+
+    # IP operations
+    def list_device_ips(self, device_id):
+        data = self.call_api("devices/%s/ips" % device_id, type="GET")
+        ips = list()
+        for jsoned in data["ip_addresses"]:
+            ip = IPAddress(jsoned)
+            ips.append(ip)
+        return ips
+
+    def get_ip(self, ip_id):
+        data = self.call_api("ips/%s" % ip_id)
+        return IPAddress(data)
+
+    def delete_ip(self, ip_id):
+        self.call_api("ips/%s" % ip_id, type="DELETE")
+
+    def list_project_ips(self, project_id, params={}):
+        data = self.call_api("projects/%s/ips" % project_id,
+                             type="GET",
+                             params=params)
+        ips = list()
+        for jsoned in data["ip_addresses"]:
+            ip = IPAddress(jsoned)
+            ips.append(ip)
+        return ips
+
+    def get_available_ip_subnets(self, ip_id, cidr):
+        data = self.call_api("/ips/%s/available" % ip_id,
+                             type="GET", params="cidr=%s" % cidr)
+        return data
+
+    def create_device_ip(self, device_id, address,
+                         manageable=True, customdata=None):
+        params = {
+            "address": address,
+            "manageable": manageable,
+            "customdata": customdata
+        }
+
+        data = self.call_api("/devices/%s/ips" % device_id,
+                             params=params, type="POST")
+        return IPAddress(data)
+
+    def reserve_ip_address(self, project_id, type, quantity, facility,
+                           details=None, comments=None, tags=list()):
+        request = {
+            "type": type,
+            "quantity": quantity,
+            "facility": facility,
+            "details": details,
+            "comments": comments,
+            "tags": tags
+        }
+
+        data = self.call_api("/projects/%s/ips" % project_id, params=request)
```

Comment:
there should be `type=POST`. Otherwise this does GET and reads something useless from the API. Even worse, the method call succeeds, and the IP is not reserved.

Reply:
Fixed